File: Scala/graph_algorithms/src/main/scala/decompositions/StronglyConnectedComponents.scala
package decompositions

object StronglyConnectedComponents {
  import scala.annotation.tailrec
  import scala.collection.mutable.ListBuffer

  type Node = Int
  type Component = List[Node]

  final case class Edge(from: Node, to: Node)
  final case class DfsOutput(components: List[Component],
                             postVisitNumbers: Vector[Node],
                             topologicalOrdering: List[Node])

  class DirectedGraph(edgeList: List[Edge], val nrNodes: Int, nodeOrder: Option[List[Node]] = None) {
    val adjacencyList: Map[Node, List[Node]] = createAdjacencyList(edgeList)
    val orderedNodes: List[Node] = getOrderedNodes(nodeOrder)

    private def getOrderedNodes(nodeOrder: Option[List[Node]]): List[Node] = nodeOrder match {
      case None => (1 to nrNodes).toList
      case Some(nodes) => nodes
    }

    private def createAdjacencyList(edgeList: List[Edge]): Map[Node, List[Node]] =
      edgeList.groupBy(_.from).mapValues(_.map(_.to))
  }

  class DepthFirstSearch(graph: DirectedGraph) {
    lazy val dfsOutput: DfsOutput = depthFirstSearch()

    private val visitStarted: Array[Int] = Array.fill(graph.nrNodes)(0)
    private val visitEnded: Array[Int] = Array.fill(graph.nrNodes)(0)
    private val topologicalSorting: ListBuffer[Node] = ListBuffer()
    private val previsitId: Iterator[Int] = Iterator.from(1)
    private val postvisitId: Iterator[Int] = Iterator.from(1)

    private def isNodeVisited(node: Node): Boolean = visitStarted(node - 1) > 0

    private def findUnvisitedNeighbour(node: Node): Option[Node] =
      graph.adjacencyList.getOrElse(node, Nil).find(!isNodeVisited(_))

    private def depthFirstSearch(): DfsOutput = {
      @tailrec
      def findComponents(nodes: List[Node], components: List[Component]): List[Component] = nodes match {
        case Nil => components
        case node :: remainingNodes =>
          if (isNodeVisited(node)) findComponents(remainingNodes, components)
          else {
            val currentComponent: Component = explore(node)
            findComponents(remainingNodes, currentComponent :: components)
          }
      }

      def explore(startingNode: Node): Component = {
        @tailrec
        def traverseComponent(previsitStack: List[Node], component: Component): Component = previsitStack match {
          case Nil => component
          case node :: restOfStack =>
            findUnvisitedNeighbour(node) match {
              case Some(neighbour) =>
                visitStarted(neighbour - 1) = previsitId.next()
                traverseComponent(neighbour :: previsitStack, neighbour :: component)
              case None =>
                visitEnded(node - 1) = postvisitId.next()
                topologicalSorting += node
                traverseComponent(restOfStack, component)
            }
        }
        visitStarted(startingNode - 1) = previsitId.next()
        traverseComponent(List(startingNode), List(startingNode))
      }

      val components: List[Component] = findComponents(graph.orderedNodes, Nil)
      DfsOutput(components, visitEnded.toVector, topologicalSorting.toList.reverse)
    }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def createGraphWithEdgesReversed(edges: List[Edge],
                                           nrNodes: Int,
                                           postVisitOrdering: Vector[Int]): DirectedGraph = {
    val reversedEdges: List[Edge] = edges.map{ case Edge(from, to) => Edge(to, from) }
    val nodeOrder: List[Node] =
      postVisitOrdering.zip(1 to nrNodes)
        .sortBy{ case (postNumber, _) => postNumber }(Ordering[Int].reverse)
        .unzip
        ._2
        .toList
    new DirectedGraph(reversedEdges, nrNodes, Some(nodeOrder))
  }

  def calcStronglyConnectedComponents(edgeList: List[Edge], nrNodes: Int): List[Component] = {
    val graph = new DirectedGraph(edgeList, nrNodes)
    val forwardDFS = new DepthFirstSearch(graph)
    val DfsOutput(_, postVisitNumbers, _) = forwardDFS.dfsOutput
    val reversedGraph: DirectedGraph = createGraphWithEdgesReversed(edgeList, nrNodes, postVisitNumbers)
    val backwardDFS = new DepthFirstSearch(reversedGraph)
    val DfsOutput(components, _, _) = backwardDFS.dfsOutput
    components
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val List(nrNodes, nrEdges): List[Int] = convertToIntList(reader.next())
    val edgeList: List[Edge] = (for { _ <- 0 until nrEdges } yield {
      val List(a, b): List[Node] = convertToIntList(reader.next())
      Edge(a, b)
    }).toList
    val scc: List[Component] = calcStronglyConnectedComponents(edgeList, nrNodes)
    println(scc.length)
  }
}

File: OCaml/graph_algorithms/decompositions/strongly_connected_components.ml
open Batteries

type edge = { left_end : int; right_end : int }
type component = int list
type dfs_output = { components : component list; postvisit_numbers : int list }

module IntDict = Hashtbl.Make(
    struct
        type t = int
        let equal = Pervasives.(=)
        let hash = Hashtbl.hash
    end)

module type DirectedGraphSig =
    sig
        type t
        val create_graph : edge list -> int -> int list option -> t
        val range: int -> int -> int list
        val depth_first_search : t -> dfs_output
    end

module DirectedGraph : DirectedGraphSig =
    struct
        type t = { adjacency_list : int list IntDict.t; nr_nodes : int; ordered_nodes : int list }

        let range a b =
            let rec loop acc left =
                if left >= b then List.rev acc
                else loop (left :: acc) (left + 1)
            in loop [] a

        let get_neighbours adjacency_list node =
            match IntDict.find_option adjacency_list node with
                | None -> []
                | Some neighbours -> neighbours

        let create_graph edges nr_nodes ordered_nodes =
            let adjacency_list = IntDict.create 100 in
            let add_edge { left_end; right_end } =
                let neighbours = get_neighbours adjacency_list left_end in
                IntDict.replace adjacency_list left_end (right_end :: neighbours) in
            List.iter add_edge edges;
            match ordered_nodes with
                | None -> { adjacency_list; nr_nodes; ordered_nodes = (range 1 (nr_nodes + 1)) }
                | Some node_ordering -> { adjacency_list; nr_nodes; ordered_nodes = node_ordering }

        let depth_first_search { adjacency_list; nr_nodes; ordered_nodes } =
            let visit_started = Array.make nr_nodes 0 in
            let visit_ended = Array.make nr_nodes 0 in
            let previsit_id = ref 1 in
            let postvisit_id = ref 1 in
            let is_node_visited node = visit_started.(node - 1) > 0 in
            let find_unvisited_neighbour node =
                List.find_opt (function elem -> not (is_node_visited elem)) (get_neighbours adjacency_list node) in
            let explore starting_node =
                let rec traverse_component previsit_stack component = match previsit_stack with
                    | [] -> component
                    | node :: rest_of_stack -> match find_unvisited_neighbour node with
                        | Some neighbour ->
                            visit_started.(neighbour - 1) <- !previsit_id;
                            previsit_id := !previsit_id + 1;
                            traverse_component (neighbour :: previsit_stack) (neighbour :: component)
                        | None ->
                            visit_ended.(node - 1) <- !postvisit_id;
                            postvisit_id := !postvisit_id + 1;
                            traverse_component rest_of_stack component in
                visit_started.(starting_node - 1) <- !previsit_id;
                previsit_id := !previsit_id + 1;
                traverse_component [starting_node] [starting_node] in
            let rec find_components nodes components = match nodes with
                | [] -> components
                | node :: remaining_nodes ->
                    if is_node_visited node then find_components remaining_nodes components
                    else
                        let current_component = explore node in
                        find_components remaining_nodes (current_component :: components) in
            let components = find_components ordered_nodes [] in
            { components; postvisit_numbers = visit_ended |> Array.to_list }
    end


let read_two_ints() = Scanf.sscanf (read_line ()) "%d %d" (fun x y -> x, y)


let read_edges nr_edges =
    let rec loop edges k =
        if k = nr_edges then edges
        else
            let left_end, right_end = read_two_ints() in
            loop ({ left_end; right_end } :: edges) (k + 1)
    in loop [] 0


let create_graph_with_edges_reversed edges nr_nodes postvisit_ordering =
    let reversed_edges = List.map (function { left_end = a; right_end = b } -> { left_end = b; right_end = a }) edges in
    let cmp (post_number1, _) (post_number2, _) = Pervasives.compare post_number2 post_number1 in
    let zipped =  List.combine postvisit_ordering (DirectedGraph.range 1 (nr_nodes + 1)) in
    let node_order = snd @@ List.split @@ List.sort cmp @@ zipped in
    DirectedGraph.create_graph reversed_edges nr_nodes (Some node_order)


let calc_strongly_connected_components edges nr_nodes =
    let graph = DirectedGraph.create_graph edges nr_nodes None in
    let { postvisit_numbers } = DirectedGraph.depth_first_search graph in
    let reversed_graph = create_graph_with_edges_reversed edges nr_nodes postvisit_numbers in
    let { components } = DirectedGraph.depth_first_search reversed_graph
    in components


let () =
    let nr_nodes, nr_edges = read_two_ints() in
    let edges = read_edges nr_edges in
    let scc = calc_strongly_connected_components edges nr_nodes in
    print_int (List.length scc);
    print_newline()
